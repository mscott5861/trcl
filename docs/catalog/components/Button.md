I'm just crazy about buttons
