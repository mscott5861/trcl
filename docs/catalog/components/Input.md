A basic Input component that maintains copies of a *real* and a *display* value (for cases in which masks are needed), and also maintains its own error state, which it reports upward to a parent **<Form/>** component.

```hint|directive
Inputs are made to be composable. Currently, additional functionality (such as validation, typeaheads, or masking) can be added by wrapping the component in the desired HOC.
```

# Props
```table
span: 6
rows:
  - Prop: activeLabel
    Required: no
    Type: string (hex/rgba/rgb)
    Description: The label displayed above the input field when active
  - Prop: activeLabelColor
    Required: no
    Type: string
    Description: The color of the activeLabel
  - Prop: bgColor
    Required: no
    Type: string (hex/rgba/rgb)
    Description: Background color of the input
  - Prop: borderless
    Required: no
    Type: boolean
    Description: If false, the input will have no border. If true, it will.
  - Prop: borderColor
    Required: no
    Type: string (hex/rgba/rgb)
    Description: Border color of the input in its default (non-errored) state
  - Prop: inputID
    Required: '**yes**'
    Type: string
    Description: A unique ID submitted by the form to identify the input's value
  - Prop: labelColor
    Required: no
    Type: string
    Description: The color of the inactive/blurred label of an empty input field
  - Prop: required
    Required: no
    Type: boolean
    Description: Whether the input contains a required field. Form will not submit with empty required fields.
```

# HOCs
