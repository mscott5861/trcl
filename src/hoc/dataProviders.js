import React from 'react'



//----------------------------------------------------------------------------------
// Swappable data provider HOCs, allowing wrapped components to consume
// {data} props for a variety of source types (files/protocols). Each
// HOC should accept a {src} prop (e.g., "ws://localhost:8080" or "file://path")
// and provide its wrapped component with a {data} prop. Currently stubbed in place
// without real testing.
//----------------------------------------------------------------------------------

export const withWSDataProvider = (WrappedComponent, src) => {
  return class extends React.Component {
    let ws = new WebSocket(this.state.src);

    constructor(props) {
      super(props);
      this.state = {
        src,
        data: []
      };
    }

    componentWillMount() {
      ws.onmessage = (e) => {
        if (e.data) {
          this.setState({
            data: e.data
          });
        }
      }

      ws.onerror = (e) => {
        console.log("Error on Websocket connection with " + this.state.src + " -> " + e.message);
      }
    }

    componentWillUnmount() {
      ws.close();
    }

    render() {
      return(
        <WrappedComponent
            data={this.state.data}
            {...this.props} />
      );
    }
  }
}

export const withHTTPDataProvider = (WrappedComponent, src) => {
  return class extends React.Component {
    constructor() {
      super();
      this.state = {
        data: [] 
      };
    }

    componentWillMount() {
      fetch(src)
        .then((response) => {
          return response.json();
        })
        .then((json) => {
          this.setState({
            data: json
          });
        });
    }

    render() {
      return(
        <WrappedComponent
          data={this.state.data}
          {...this.props}/>
      );
    }
  }
}

