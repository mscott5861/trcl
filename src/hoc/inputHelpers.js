import React from 'react'
import styled from 'styled-components'
import { isRequired } from 'utilities'



//----------------------------------------------------------------------------------
// Swappable input helper HOCs for standard input functionality (typeahead, vali-
// dation, error state handling, disabling/enabling, etc.)
//----------------------------------------------------------------------------------
export const withMask = (WrappedInput, mask = isRequired('mask is a required parameter for the withMask HOC.')) => {
  return class extends React.Component {
    maskInput = (inputReceived) => {
      let displayValue = '';

      for (let i = 0; i < inputReceived.length; i++) {
        displayValue += mask;
      }

      
      return displayValue;
    }

    render() {
      return (
        <WrappedInput
          maskInput={this.maskInput}
          {...this.props} />
      );
    }
  }
}


// TODO: function stub
export const withTypeAhead = (WrappedInput, value) => {
  return class extends React.Component {
    render() {
      return (
        <WrappedInput
            value={this.state.value}
            {...this.props} />
      );
    }
  }
}


const StStatusBlock = styled.div`
  position: absolute;
  right: 0;
  height: 100%;
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  width: 35px;
  background: ${props => props.valid === null || props.inputIsEmpty ? 'transparent' : props.valid ? 'linear-gradient(to right, rgba(0, 255, 0, 0), rgba(0, 255, 0, 0.1))' : 'linear-gradient(to right, rgba(255,0,0,0), rgba(255,0,0,0.1))'};

linear-gradient(to right, rgba(255,0,0,0), rgba(255,0,0,1))
  transition: background .15s linear;
`

const StValidationIcon = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  height: 100%;
  width: 100%;
  padding: 0.5rem;

  & span {
    position: absolute;
    height: 1px;
    width: calc(100% - 1rem);
    transition: all .15s linear;
    box-sizing: border-box;
    border-radius: 3px;
  }

  & span:nth-child(1) {
    top: ${props => props.status === 'empty' ? '50%' : 
                    props.status === 'hasError' ? 'initial' :
                    '55%'};
    left: ${props => props.status === 'empty' ? '50%' : 'initial'};
    transform: ${props => props.status === 'empty' ? 'translateY(-50%) translateX(-50%)' : 'rotate(45deg)'};
    background-color: ${props => props.status === 'empty' ? '#777' :
                                 props.status === 'hasError' ? '#C45256' :
                                 '#55C452'};
    width: ${props => props.status === 'empty' ? '0%' :
             (props.status === 'valid' ? '0.55rem' : 'calc(100% - 1rem)')};
  }

  & span:nth-child(2) {
    background-color: ${props => props.status === 'empty' ? '#777' :
                                 props.status === 'hasError' ? '#C45256' :
                                 '#55C452'};
    top: initial;
    width: ${props => props.status === 'empty' ? '0' : props.status === 'valid' && '1rem'};
    left: ${props => props.status === 'valid' && '0.75rem'};
    transform: ${props => props.status === 'empty' ? 'rotate(90deg)' : 
                 props => props.status === 'hasError' ? 'rotate(-45deg)' :
                 'rotate(-45deg)'};
    }
  }

`

const ValidationIcon = function(props) {
  return (
    <StValidationIcon
      status={props.status}>
      <span/>
      <span/> 
    </StValidationIcon>
  );
}


export const withValidation = (WrappedInput, schemaPackage = isRequired('schemaPackage is a required object for the withValidation HOC.')) => {
  return class extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        valid: null,
        inputIsEmpty: true,
        errorMessage: '',
      }
    }

    validateInput = (inputReceived) => {
      let regex = new RegExp(schemaPackage.schema),
          valid = regex.test(inputReceived);

      this.setState({
        valid,
        inputIsEmpty: (inputReceived.length === 0),
        errorMessage: (!valid ? schemaPackage.errorMessage : ''),
      });

      return inputReceived;
    }

    render() {
      return (
        <WrappedInput
          errorMessage={this.state.errorMessage}
          validateInput={this.validateInput}
          {...this.props}>
          <StStatusBlock
            valid={this.state.valid}
            inputIsEmpty={this.state.inputIsEmpty}>
            <ValidationIcon
              status={this.state.inputIsEmpty ? 'empty' : (this.state.errorMessage && this.state.errorMessage.length > 0 ?
                      'hasError' : 'valid')}/>
          </StStatusBlock>
        </WrappedInput>
      );
    }
  }
}
