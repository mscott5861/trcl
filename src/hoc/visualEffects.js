import React from 'react'
import styled from 'styled-components'

const ParallaxGroup = styled.div`
  position: relative; 
  transform-style: preserve-3d;
  height: 100vh;
    
  & :nth-child(1) {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    overflow-y: ${props => props.pageOrientation === 'portrait' ? 'scroll' : 'initial'};
    overflow-x: ${props => props.pageOrientation === 'portrait' ? 'initial' : 'scroll'};
    transform: ${props => props.depth ? 'translateZ(' + props.depth + 'px) scale(' + (1 + (props.depth * -1)) + ') !important' :
                'translateZ(-0.5px) scale(1.5) !important'};
    z-index: -10;
  }
}`


export const withParallax = (WrappedComponent, depth, pageOrientation = 'portrait') => {
  return class extends React.Component {
    constructor() {
      super();
      let style = document.createElement('style');
      style.type = 'text/css';
      style.id = 'withParallaxHOC';
      style.innerHTML = `.plx { 
        perspective: 1px;
        height: 100vh;
        overflow-x: ` + (pageOrientation === 'portrait' ? 'hidden;' : 'scroll;') +
       `overflow-y: ` + (pageOrientation === 'portrait' ? 'scroll;' : 'hidden;') +
       `}
      html {
        overflow: hidden;
      }`;
      
      let styleSheet = document.getElementById('withParallaxHOC');
      styleSheet === null && document.getElementsByTagName('head')[0].appendChild(style);
    }

    componentWillMount() {
      document.body.classList.add('plx');
    }

    componentWillUnmount() {
      document.body.classList.remove('plx');
    }

    render() {
      return(
        <ParallaxGroup
          depth={-1 * depth}
          pageOrintation={pageOrientation}>
          <WrappedComponent
            {...this.props} />
        </ParallaxGroup>
      );
    }
  }
}

const StRelativeContainer = styled.div`
  position: relative;
`

const StDuotoneContainer = styled.div`
  &::before,
  &::after {
    position: absolute;
    display: block;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    content: '';
  }

  &::before {
    background-color: #F00E2E;
    mix-blend-mode: darken;
  }

  &::after {
    background-color: #192550;
    mix-blend-mode: lighten;
  }
`

export const withDuotone = (WrappedComponent) => {
  return class extends React.Component {
    render() {
      return(
        <StRelativeContainer>
          <StDuotoneContainer>
            <WrappedComponent
              {...this.props} />
          </StDuotoneContainer>
        </StRelativeContainer>
      );
    }
  }
}
