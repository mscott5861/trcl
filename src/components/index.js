import CodeBlock from './atomic/containers/CodeBlock'
import Copy from './atomic/text/Copy'
import H1 from './atomic/text/H1'
import H2 from './atomic/text/H2'
import H3 from './atomic/text/H3'
import HR from './atomic/embellishments/HR'
import HeroBanner from './atomic/images/HeroBanner'
import ImageWithCaption from './atomic/images/ImageWithCaption'
import Input from './atomic/text/Input'
import Leaflet from './atomic/containers/Leaflet'
import Link from './atomic/text/Link'
import MainContent from './atomic/containers/MainContent'
import Navbar from './atomic/containers/Navbar'
import PropsTable from './composite/containers/PropsTable'
import Sidebar from './atomic/containers/Sidebar'
import SingleColumnWithSidebar from './atomic/layouts/page-layouts/SingleColumnWithSidebar'
import SingleColumnWithSidebarAndNavbar from './atomic/layouts/page-layouts/SingleColumnWithSidebarAndNavbar'
import Table from './atomic/containers/Table'
import ThreeColumnLayout from './atomic/layouts/sublayouts/ThreeColumnLayout'
import TwoColumnLayout from './atomic/layouts/sublayouts/TwoColumnLayout'

export {
  CodeBlock,
  Copy,
  H1,
  H2,
  H3,
  HR,
  HeroBanner,
  ImageWithCaption,
  Input,
  Leaflet,
  Link,
  MainContent,
  Navbar,
  PropsTable,
  Sidebar,
  SingleColumnWithSidebar,
  SingleColumnWithSidebarAndNavbar,
  Table,
  ThreeColumnLayout,
  TwoColumnLayout,
}
