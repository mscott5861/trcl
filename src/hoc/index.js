import { withHTTPDataProvider } from './dataProviders'
import { withMask } from './inputHelpers'
import { withParallax } from './visualEffects'
import { withSubmitHandler } from './dataProviders'
import { withValidation } from './inputHelpers'
import { withWSDataProvider } from './dataProviders'

export {
    withHTTPDataProvider,
    withMask,
    withParallax,
    withSubmitHandler,
    withValidation,
    withWSDataProvider,
};
