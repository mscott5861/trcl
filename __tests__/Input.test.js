import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import Enzyme, { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import { Input } from 'components'
import { withMask, withValidation } from 'hoc'

Enzyme.configure({
  adapter: new Adapter()
});


//------------------------------------------------------------------
// Snapshot the Input
//------------------------------------------------------------------
describe('Input', () => {
  it('renders correctly', () => {
    const output = shallow(
      <Input
        activeLabel='test activeLabel'
        activeLabelColor='#000'
        bgColor='#fff'
        borderless={false}
        borderColor='red'
        inputID='test inputID'
        labelColor='green'
        required
        />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});



//------------------------------------------------------------------
// Ensure displayValue is updated on keystroke
//------------------------------------------------------------------
it('updates displayValue when given keystroke', () => {
  const output = shallow(<Input/>),
        testVal = 'Z';

  expect(output.state().displayValue).toEqual('');
  output.simulate('change', {
    target: {
      name: 'test',
      value: testVal
    }
  });
  expect(output.state().displayValue).toEqual(testVal);
});



//------------------------------------------------------------------
// Ensure realValue is updated on keystroke
//------------------------------------------------------------------
it('updates realValue when given keystroke', () => {
  const output = shallow(<Input/>),
        testVal = 'Z';

  expect(output.state().realValue).toEqual('');
  output.simulate('change', {
    target: {
      name: 'test',
      value: testVal
    }
  });
  expect(output.state().realValue).toEqual(testVal);
});



//------------------------------------------------------------------
// Required components should error (hasError = true) on creation
//------------------------------------------------------------------
it('state.hasError=true on creation when props.required=true', () => {
  const output = shallow(<Input required/>);
  expect(output.state().hasError).toEqual(true);
});



//------------------------------------------------------------------
// Unrequired components should not error (hasError = true) on 
// creation
//------------------------------------------------------------------
it('state.hasError=false on creation when props.required=false', () => {
  const output = shallow(<Input/>);
  expect(output.state().hasError).toEqual(false);
});



//------------------------------------------------------------------
// Snapshot the MaskedInput
//------------------------------------------------------------------
describe('MaskedInput', () => {
  it('renders correctly', () => {
    const MaskedInput = withMask(Input, '*');
    const output = shallow(
        <MaskedInput
          activeLabel='test activeLabel'
          activeLabelColor='#000'
          bgColor='#fff'
          borderless={false}
          borderColor='red'
          inputID='test inputID'
          labelColor='green'
          required
          />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});



//------------------------------------------------------------------
// Inputs wrapped in withMask HOC should report the passed mask
// as the displayValue on keystroke
//------------------------------------------------------------------
it('updates state.displayValue with mask when given keystroke and wrapped in withMask HOC', () => {
  const MaskedInput = withMask(Input, '*'),
        maskedInput = shallow(<MaskedInput/>),
        testVal = 'Z';

  const input = maskedInput.find('Input').shallow();

  expect(input.state().displayValue).toEqual('');
  input.simulate('change', {
    target: {
      name: 'test',
      value: testVal,
    }}
  );

  expect(input.state().displayValue).toEqual('*');
});



//------------------------------------------------------------------
// Snapshot the ValidatedInput
//------------------------------------------------------------------
const atLeastThreeSchema = {
  schema: "(.*[a-zA-Z0-9]){3}",
  errorMessage: "A test error"
};

describe('ValidatedInput', () => {
  it('renders correctly', () => {
    const ValidatedInput = withValidation(Input, atLeastThreeSchema);
    const output = shallow(
        <ValidatedInput
          activeLabel='test activeLabel'
          activeLabelColor='#000'
          bgColor='#fff'
          borderless={false}
          borderColor='red'
          inputID='test inputID'
          labelColor='green'
          required
          />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});



//------------------------------------------------------------------
// Inputs wrapped in withValidation HOC should error when failing
// their validation schema
//------------------------------------------------------------------
it('errors when wrapped in withValidation HOC but input fails the schema', () => {
  const ValidatedInput = withValidation(Input, atLeastThreeSchema),
        validatedInput = shallow(<ValidatedInput/>),
        testVal = 'Z';

  const input = validatedInput.find('Input').shallow();

  expect(input.props().errorMessage).toEqual('');
  input.simulate('change', {
    target: {
      name: 'test',
      value: testVal,
    }}
  );

  // TODO: See if Jest/Enzyme offers a way of ensuring the functions
  // it simulates actually finish execution (without using timeouts).
  setTimeout(() => {
    expect(input.props().errorMessage).toEqual('A test error');
  }, 500);
});
