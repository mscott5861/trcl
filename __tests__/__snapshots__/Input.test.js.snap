// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders correctly 1`] = `
<styled.div
  bgColor="#fff"
  borderColor="red"
  borderless={false}
  displayValue=""
  isFocused={false}
  onChange={[Function]}
>
  <styled.input
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  />
  <styled.p
    isFocused={false}
    labelColor="green"
  />
  <styled.p
    activeLabelColor="#000"
    isFocused={false}
    required={true}
  >
    test activeLabel
  </styled.p>
</styled.div>
`;

exports[`MaskedInput renders correctly 1`] = `
<Input
  activeLabel="test activeLabel"
  activeLabelColor="#000"
  bgColor="#fff"
  borderColor="red"
  borderless={false}
  inputID="test inputID"
  labelColor="green"
  maskInput={[Function]}
  required={true}
/>
`;

exports[`ValidatedInput renders correctly 1`] = `
<Input
  activeLabel="test activeLabel"
  activeLabelColor="#000"
  bgColor="#fff"
  borderColor="red"
  borderless={false}
  errorMessage=""
  inputID="test inputID"
  labelColor="green"
  required={true}
  validateInput={[Function]}
>
  <styled.div
    inputIsEmpty={true}
    valid={null}
  >
    <ValidationIcon
      status="empty"
    />
  </styled.div>
</Input>
`;
